server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,225 @@
+			registeredApp.Manifest = manifest
+			if err := adm.store.StoreApp(registeredApp); err != nil {
+				return errors.Wrapf(err, "can't store app - %s", registeredApp.Manifest.AppID)
+			}
+			updatedAppVersionsMap[registeredApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as registered
+	for appID, manifest := range adm.store.ListManifests() {
+		if _, ok := registeredAppsMap[appID]; ok {
+			continue
+		}
+		if err := adm.AddApp(manifest); err != nil {
+			return errors.Wrapf(err, "can't add new app as registered appID - %s", manifest.AppID)
+		}
+	}
+
+	registeredAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
+	for _, registeredApp := range registeredAppsUpgraded {
+		if registeredApp.Status == apps.AppStatusEnabled {
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[registeredApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[registeredApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app. It should be called only once
+			f := func() error {
+				if err := adm.call(registeredApp, registeredApp.Manifest.OnStartup, values); err != nil {
+					adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+				}
+				return nil
+			}
+			if err := adm.callOnce(f); err != nil {
+				adm.mm.Log.Error("Can't callOnce the onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *apps.App) error {
+	// Call delete the function of the app
+	if err := adm.call(app, app.Manifest.OnDelete, nil); err != nil {
+		return errors.Wrapf(err, "delete failed. appID - %s", app.Manifest.AppID)
+	}
+
+	// delete oauth app
+	conf := adm.conf.GetConfig()
+	client := model.NewAPIv4Client(conf.MattermostSiteURL)
+	client.SetToken(string(adm.adminToken))
+
+	if app.OAuth2ClientID != "" {
+		success, response := client.DeleteOAuthApp(app.OAuth2ClientID)
+		if !success || response.StatusCode != http.StatusNoContent {
+			return errors.Wrapf(response.Error, "failed to delete OAuth2 App - %s", app.Manifest.AppID)
+		}
+	}
+
+	// delete app from proxy plugin, not removing the data
+	if err := adm.store.DeleteApp(app); err != nil {
+		return errors.Wrapf(err, "can't delete app - %s", app.Manifest.AppID)
+	}
+
+	adm.mm.Log.Info("Deleted the app", "app_id", app.Manifest.AppID)
+
+	return nil
+}
+
+func (adm *Admin) AddApp(manifest *apps.Manifest) error {
+	newApp := &apps.App{}
+	newApp.Manifest = manifest
+	newApp.Status = apps.AppStatusRegistered
+	if err := adm.store.StoreApp(newApp); err != nil {
+		return errors.Wrapf(err, "can't store app - %s", manifest.AppID)
+	}
+	adm.mm.Log.Info("App is registered", "app_id", manifest.AppID)
+	return nil
+}
+
+func (adm *Admin) callOnce(f func() error) error {
+	// Delete previous job
+	key := "PP_CallOnce_key"
```

Comment:
1. 2/5 I do not find the prefix `PP` to be intention revealing.
2. 0/5 All key names should be const, and (somewhat) consolidated, declared in 1 (or a few) place(s)

Reply:
PP = proxy plugin. Why not?

Reply:
I work on it, and I have to think about what it means every time. Why do we need this acronym? The keys are already in the namespace of the plugin.